File: apps/macos_native/Sources/PasswordManagerMacOSApp/Sync/EncryptedRemoteSyncClient.swift
import Foundation

final class EncryptedRemoteSyncClient: RemoteSyncClient, @unchecked Sendable {
    private let delegate: RemoteSyncClient
    private let crypto: VaultCryptoService
    private let vaultKey: Data
    private let masterKeyRecord: MasterKeyRecord?
    private let includeMasterKeyRecord: Bool
    private(set) var downloadedPlaintextRemote = false

    init(
        delegate: RemoteSyncClient,
        crypto: VaultCryptoService,
        vaultKey: Data,
        masterKeyRecord: MasterKeyRecord?,
        includeMasterKeyRecord: Bool
    ) {
        self.delegate = delegate
        self.crypto = crypto
        self.vaultKey = vaultKey
        self.masterKeyRecord = masterKeyRecord
        self.includeMasterKeyRecord = includeMasterKeyRecord
    }

    func metadata() async -> RemoteSyncMetadata {
        let metadata = await delegate.metadata()
        return RemoteSyncMetadata(statusCode: metadata.statusCode)
    }

    func download() async -> RemoteSyncResult {
        let download = await delegate.download()
        guard Self.isSuccessfulDownload(download.statusCode),
              let rawPayload = download.payload,
              !rawPayload.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty,
              Self.containsEncryptedVault(rawPayload) else {
            markPlaintextRemoteIfNeeded(download.payload)
            return download
        }
        do {
            let payload = try decodeEncryptedPayload(rawPayload)
            return RemoteSyncResult(payload: payload, statusCode: download.statusCode)
        } catch {
            return RemoteSyncResult(payload: nil, statusCode: 422)
        }
    }

    private func markPlaintextRemoteIfNeeded(_ rawPayload: String?) {
        guard let rawPayload,
              let data = rawPayload.data(using: .utf8),
              (try? Self.makeDecoder().decode(VaultSyncPayload.self, from: data)) != nil else {
            return
        }
        downloadedPlaintextRemote = true
    }

    func upload(_ payload: String) async -> RemoteSyncResult {
        do {
            let encryptedPayload = try encodeEncryptedPayload(payload)
            return await delegate.upload(encryptedPayload)
        } catch {
            return RemoteSyncResult(payload: nil, statusCode: 422)
        }
    }

    private func encodeEncryptedPayload(_ payload: String) throws -> String {
        let decoder = Self.makeDecoder()
        let encoder = Self.makeEncoder()
        guard let data = payload.data(using: .utf8) else {
            throw VaultSyncEngineError.invalidRemotePayload
        }
        let syncPayload = try decoder.decode(VaultSyncPayload.self, from: data)
        let encryptedVault = try crypto.encrypt(try encoder.encode(syncPayload.snapshot), key: vaultKey)
        let envelope = EncryptedVaultSyncEnvelope(
            version: syncPayload.version,
            exportedAt: syncPayload.exportedAt,
            deviceId: syncPayload.deviceId,
            revision: syncPayload.revision,
            masterKeyRecord: includeMasterKeyRecord ? masterKeyRecord : nil,
            encryptedVault: encryptedVault
        )
        return String(data: try encoder.encode(envelope), encoding: .utf8) ?? "{}"
    }

    private func decodeEncryptedPayload(_ rawPayload: String) throws -> String {
        let decoder = Self.makeDecoder()
        let encoder = Self.makeEncoder()
        guard let data = rawPayload.data(using: .utf8) else {
            throw VaultSyncEngineError.invalidRemotePayload
        }
        let envelope = try decoder.decode(EncryptedVaultSyncEnvelope.self, from: data)
        let snapshot = try decoder.decode(VaultSnapshot.self, from: crypto.decrypt(envelope.encryptedVault, key: vaultKey))
        let payload = VaultSyncPayload(
            version: envelope.version,
            exportedAt: envelope.exportedAt,
            deviceId: envelope.deviceId,
            revision: envelope.revision,
            snapshot: snapshot
        )
        return String(data: try encoder.encode(payload), encoding: .utf8) ?? "{}"
    }

    private static func containsEncryptedVault(_ rawPayload: String) -> Bool {
        guard let data = rawPayload.data(using: .utf8),
              let object = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            return false
        }
        return object["encryptedVault"] != nil
    }

    private static func isSuccessfulDownload(_ statusCode: Int) -> Bool {
        (statusCode >= 200 && statusCode < 300) || statusCode == 404
    }

    private static func makeEncoder() -> JSONEncoder {
        let encoder = JSONEncoder()
        encoder.outputFormatting = [.sortedKeys]
        encoder.dateEncodingStrategy = .iso8601
        return encoder
    }

    private static func makeDecoder() -> JSONDecoder {
        let decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .iso8601
        return decoder
    }
}

private struct EncryptedVaultSyncEnvelope: Codable, Sendable {
    var version: Int = 1
    var exportedAt: Date
    var deviceId: String
    var revision: Int
    var masterKeyRecord: MasterKeyRecord?
    var encryptedVault: EncryptedPayloadRecord
}

File: apps/android_native/app/src/main/java/life/devops/passwordmanager/sync/EncryptedRemoteSyncClient.kt
package life.devops.passwordmanager.sync

import life.devops.passwordmanager.model.EncryptedPayloadRecord
import life.devops.passwordmanager.model.MasterKeyRecord
import life.devops.passwordmanager.store.AndroidVaultCrypto
import life.devops.passwordmanager.store.VaultJson
import org.json.JSONObject
import org.json.JSONObject.NULL
import java.nio.charset.StandardCharsets
import java.time.Instant

class EncryptedRemoteSyncClient(
    private val delegate: RemoteSyncClient,
    private val crypto: AndroidVaultCrypto,
    vaultKey: ByteArray,
    masterKeyRecord: MasterKeyRecord?,
    private val masterPassword: String,
    private val includeMasterKeyRecord: Boolean,
) : RemoteSyncClient {
    private var effectiveVaultKey: ByteArray = vaultKey.copyOf()
    private var effectiveMasterKeyRecord: MasterKeyRecord? = masterKeyRecord
    private var remoteVaultKey: ByteArray? = null

    var downloadedPlaintextRemote: Boolean = false
        private set
    var remoteMasterKeyRecord: MasterKeyRecord? = null
        private set

    fun resolvedRemoteVaultKey(): ByteArray? = remoteVaultKey?.copyOf()

    override fun metadata(): RemoteSyncMetadata =
        delegate.metadata().let { metadata ->
            metadata.copy(eTag = null, lastModified = null, contentLength = null)
        }

    override fun download(): RemoteSyncResult {
        val download = delegate.download()
        val rawPayload = download.payload?.trim()
        if (!isSuccessfulDownload(download.statusCode) || rawPayload.isNullOrEmpty()) {
            return download
        }
        val encryptedPayload = decodeEncryptedPayload(rawPayload)
        if (encryptedPayload == null) {
            markPlaintextRemoteIfNeeded(rawPayload)
            return download
        }
        return download.copy(payload = encryptedPayload.toJson().toString())
    }

    override fun upload(payload: String): RemoteSyncResult {
        val syncPayload = VaultSyncPayload.fromJson(payload)
        val encryptedVault = crypto.encrypt(
            VaultJson.encodeSnapshot(syncPayload.snapshot).toByteArray(StandardCharsets.UTF_8),
            effectiveVaultKey,
        )
        val envelope = JSONObject()
            .put("version", syncPayload.version)
            .put("exportedAt", syncPayload.exportedAt.toString())
            .put("deviceId", syncPayload.deviceId)
            .put("revision", syncPayload.revision)
            .put(
                "masterKeyRecord",
                if (includeMasterKeyRecord) effectiveMasterKeyRecord?.toJson() ?: NULL else NULL,
            )
            .put("encryptedVault", encryptedVault.toJson())
        return delegate.upload(envelope.toString())
    }

    private fun decodeEncryptedPayload(rawPayload: String): VaultSyncPayload? {
        val json = JSONObject(rawPayload)
        val encryptedVault = json.optJSONObject("encryptedVault")?.toEncryptedPayloadRecord()
            ?: return null
        val decryptedVault = decryptWithAvailableKey(json, encryptedVault)
        val decryptedSnapshot = VaultJson.decodeSnapshot(
            String(decryptedVault, StandardCharsets.UTF_8)
        )
        return VaultSyncPayload(
            version = json.optInt("version", 1),
            exportedAt = json.optInstant("exportedAt") ?: Instant.EPOCH,
            deviceId = json.optString("deviceId"),
            revision = json.optInt("revision", 0),
            snapshot = decryptedSnapshot,
        )
    }

    private fun decryptWithAvailableKey(
        envelope: JSONObject,
        encryptedVault: EncryptedPayloadRecord,
    ): ByteArray = try {
        crypto.decrypt(encryptedVault, effectiveVaultKey)
    } catch (localKeyError: Exception) {
        val remoteRecord = envelope.optJSONObject("masterKeyRecord")?.toMasterKeyRecord()
            ?: throw localKeyError
        val resolvedKey = crypto.verify(masterPassword, remoteRecord)
        val decrypted = crypto.decrypt(encryptedVault, resolvedKey)
        effectiveVaultKey = resolvedKey.copyOf()
        effectiveMasterKeyRecord = remoteRecord
        remoteVaultKey = resolvedKey.copyOf()
        remoteMasterKeyRecord = remoteRecord
        decrypted
    }

    private fun markPlaintextRemoteIfNeeded(rawPayload: String) {
        runCatching { VaultSyncPayload.fromJson(rawPayload) }
            .onSuccess { downloadedPlaintextRemote = true }
    }

    private fun MasterKeyRecord.toJson(): JSONObject =
        JSONObject()
            .put("saltBase64", saltBase64)
            .put("iterations", iterations)
            .put("verifierBase64", verifierBase64)
            .put("metadataSaltBase64", metadataSaltBase64 ?: NULL)
            .put("metadataIterations", metadataIterations ?: NULL)

    private fun EncryptedPayloadRecord.toJson(): JSONObject =
        JSONObject()
            .put("ciphertext", ciphertext)
            .put("nonce", nonce)
            .put("mac", mac)
            .put("version", version)

    private fun JSONObject.toMasterKeyRecord(): MasterKeyRecord =
        MasterKeyRecord(
            saltBase64 = getString("saltBase64"),
            iterations = getInt("iterations"),
            verifierBase64 = getString("verifierBase64"),
            metadataSaltBase64 = optionalString("metadataSaltBase64"),
            metadataIterations = optionalInt("metadataIterations"),
        )

    private fun JSONObject.toEncryptedPayloadRecord(): EncryptedPayloadRecord =
        EncryptedPayloadRecord(
            ciphertext = optString("ciphertext"),
            nonce = optString("nonce"),
            mac = optString("mac"),
            version = optInt("version", 1),
        )

    private fun JSONObject.optInstant(name: String): Instant? {
        if (!has(name) || isNull(name)) return null
        return runCatching { Instant.parse(optString(name)) }.getOrNull()
    }

    private fun JSONObject.optionalString(name: String): String? {
        if (!has(name) || isNull(name)) return null
        return optString(name).takeIf { it.isNotEmpty() }
    }

    private fun JSONObject.optionalInt(name: String): Int? {
        if (!has(name) || isNull(name)) return null
        return optInt(name)
    }

    private fun isSuccessfulDownload(statusCode: Int): Boolean =
        statusCode in 200..299 || statusCode == 404
}
